Fail to install Matlab Wrapper

## Description

Following the github instructions I get the following error installing the matlab toolbox at the 100% mark.

## Steps to reproduce

1. Follow github instructions:
```
#!bash
mkdir build
cd build
cmake -DCMAKE_PREFIX_PATH=/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/boost_1_87_0 -DCMAKE_INSTALL_PREFIX=/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam_install -DGTSAM_INSTALL_MATLAB_TOOLBOX=ON ..
make check (optional, runs unit tests)
make install
```
make intstall - Error:

```
[100%] Generating ../wrap/gtsam/gtsam_wrapper.cpp
[MatlabWrapper] Ignoring classes: ['gtsam::Point2', 'gtsam::Point3', 'gtsam::CustomFactor']
[100%] Building CXX object matlab/CMakeFiles/gtsam_matlab_wrapper.dir/__/wrap/gtsam/gtsam_wrapper.cpp.o
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build/wrap/gtsam/gtsam_wrapper.cpp: In function ‘void gtsamDiscreteSearch_FromFactorGraph_1013(int, mxArray**, int, const mxArray**)’:
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build/wrap/gtsam/gtsam_wrapper.cpp:15444:45: error: ‘DiscreteSearch’ was not declared in this scope; did you mean ‘gtsam::DiscreteSearch’?
15444 |   out[0] = wrap_shared_ptr(std::make_shared<DiscreteSearch>(gtsam::DiscreteSearch::FromFactorGraph(factorGraph,ordering,buildJunctionTree)),"DiscreteSearch", false);
      |                                             ^~~~~~~~~~~~~~
      |                                             gtsam::DiscreteSearch
In file included from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build/wrap/gtsam/gtsam_wrapper.cpp:31:
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteSearch.h:44:20: note: ‘gtsam::DiscreteSearch’ declared here
   44 | class GTSAM_EXPORT DiscreteSearch {
      |                    ^~~~~~~~~~~~~~
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build/wrap/gtsam/gtsam_wrapper.cpp:15444:139: error: no matching function for call to ‘make_shared<<expression error> >(gtsam::DiscreteSearch)’
15444 | hared<DiscreteSearch>(gtsam::DiscreteSearch::FromFactorGraph(factorGraph,ordering,buildJunctionTree)),"DiscreteSearch", false);
      |                                                                                                     ^

In file included from /usr/include/c++/9/memory:81,
                 from /usr/include/tbb/tbb_stddef.h:452,
                 from /usr/include/tbb/scalable_allocator.h:204,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/base/types.h:31,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/global_includes.h:22,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/base/Vector.h:29,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/wrap/matlab.h:25,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build/wrap/gtsam/gtsam_wrapper.cpp:1:
/usr/include/c++/9/bits/shared_ptr.h:714:5: note: candidate: ‘template<class _Tp, class ... _Args> std::shared_ptr<_Tp> std::make_shared(_Args&& ...)’
  714 |     make_shared(_Args&&... __args)
      |     ^~~~~~~~~~~
/usr/include/c++/9/bits/shared_ptr.h:714:5: note:   template argument deduction/substitution failed:
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build/wrap/gtsam/gtsam_wrapper.cpp:15444:139: error: template argument 1 is invalid
15444 | hared<DiscreteSearch>(gtsam::DiscreteSearch::FromFactorGraph(factorGraph,ordering,buildJunctionTree)),"DiscreteSearch", false);
      |                                                                                                     ^

/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build/wrap/gtsam/gtsam_wrapper.cpp: In function ‘void gtsamDiscreteSearch_FromFactorGraph_1014(int, mxArray**, int, const mxArray**)’:
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build/wrap/gtsam/gtsam_wrapper.cpp:15452:45: error: ‘DiscreteSearch’ was not declared in this scope; did you mean ‘gtsam::DiscreteSearch’?
15452 |   out[0] = wrap_shared_ptr(std::make_shared<DiscreteSearch>(gtsam::DiscreteSearch::FromFactorGraph(factorGraph,ordering,false)),"DiscreteSearch", false);
      |                                             ^~~~~~~~~~~~~~
      |                                             gtsam::DiscreteSearch
In file included from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build/wrap/gtsam/gtsam_wrapper.cpp:31:
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteSearch.h:44:20: note: ‘gtsam::DiscreteSearch’ declared here
   44 | class GTSAM_EXPORT DiscreteSearch {
      |                    ^~~~~~~~~~~~~~
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build/wrap/gtsam/gtsam_wrapper.cpp:15452:127: error: no matching function for call to ‘make_shared<<expression error> >(gtsam::DiscreteSearch)’
15452 | (std::make_shared<DiscreteSearch>(gtsam::DiscreteSearch::FromFactorGraph(factorGraph,ordering,false)),"DiscreteSearch", false);
      |                                                                                                     ^

In file included from /usr/include/c++/9/memory:81,
                 from /usr/include/tbb/tbb_stddef.h:452,
                 from /usr/include/tbb/scalable_allocator.h:204,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/base/types.h:31,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/global_includes.h:22,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/base/Vector.h:29,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/wrap/matlab.h:25,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build/wrap/gtsam/gtsam_wrapper.cpp:1:
/usr/include/c++/9/bits/shared_ptr.h:714:5: note: candidate: ‘template<class _Tp, class ... _Args> std::shared_ptr<_Tp> std::make_shared(_Args&& ...)’
  714 |     make_shared(_Args&&... __args)
      |     ^~~~~~~~~~~
/usr/include/c++/9/bits/shared_ptr.h:714:5: note:   template argument deduction/substitution failed:
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build/wrap/gtsam/gtsam_wrapper.cpp:15452:127: error: template argument 1 is invalid
15452 | (std::make_shared<DiscreteSearch>(gtsam::DiscreteSearch::FromFactorGraph(factorGraph,ordering,false)),"DiscreteSearch", false);
      |                                                                                                     ^

In file included from /usr/include/x86_64-linux-gnu/c++/9/bits/c++allocator.h:33,
                 from /usr/include/c++/9/bits/allocator.h:46,
                 from /usr/include/c++/9/string:41,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/base/types.h:28,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/global_includes.h:22,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/base/Vector.h:29,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/wrap/matlab.h:25,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build/wrap/gtsam/gtsam_wrapper.cpp:1:
/usr/include/c++/9/ext/new_allocator.h: In instantiation of ‘void __gnu_cxx::new_allocator<_Tp>::construct(_Up*, _Args&& ...) [with _Up = gtsam::DiscreteConditional; _Args = {std::shared_ptr<gtsam::DiscreteConditional>}; _Tp = gtsam::DiscreteConditional]’:
/usr/include/c++/9/bits/alloc_traits.h:483:4:   required from ‘static void std::allocator_traits<std::allocator<_CharT> >::construct(std::allocator_traits<std::allocator<_CharT> >::allocator_type&, _Up*, _Args&& ...) [with _Up = gtsam::DiscreteConditional; _Args = {std::shared_ptr<gtsam::DiscreteConditional>}; _Tp = gtsam::DiscreteConditional; std::allocator_traits<std::allocator<_CharT> >::allocator_type = std::allocator<gtsam::DiscreteConditional>]’
/usr/include/c++/9/bits/shared_ptr_base.h:548:39:   required from ‘std::_Sp_counted_ptr_inplace<_Tp, _Alloc, _Lp>::_Sp_counted_ptr_inplace(_Alloc, _Args&& ...) [with _Args = {std::shared_ptr<gtsam::DiscreteConditional>}; _Tp = gtsam::DiscreteConditional; _Alloc = std::allocator<gtsam::DiscreteConditional>; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/9/bits/shared_ptr_base.h:679:16:   required from ‘std::__shared_count<_Lp>::__shared_count(_Tp*&, std::_Sp_alloc_shared_tag<_Alloc>, _Args&& ...) [with _Tp = gtsam::DiscreteConditional; _Alloc = std::allocator<gtsam::DiscreteConditional>; _Args = {std::shared_ptr<gtsam::DiscreteConditional>}; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/9/bits/shared_ptr_base.h:1344:71:   required from ‘std::__shared_ptr<_Tp, _Lp>::__shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<gtsam::DiscreteConditional>; _Args = {std::shared_ptr<gtsam::DiscreteConditional>}; _Tp = gtsam::DiscreteConditional; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/9/bits/shared_ptr.h:359:59:   required from ‘std::shared_ptr<_Tp>::shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<gtsam::DiscreteConditional>; _Args = {std::shared_ptr<gtsam::DiscreteConditional>}; _Tp = gtsam::DiscreteConditional]’
/usr/include/c++/9/bits/shared_ptr.h:701:14:   required from ‘std::shared_ptr<_Tp> std::allocate_shared(const _Alloc&, _Args&& ...) [with _Tp = gtsam::DiscreteConditional; _Alloc = std::allocator<gtsam::DiscreteConditional>; _Args = {std::shared_ptr<gtsam::DiscreteConditional>}]’
/usr/include/c++/9/bits/shared_ptr.h:717:39:   required from ‘std::shared_ptr<_Tp> std::make_shared(_Args&& ...) [with _Tp = gtsam::DiscreteConditional; _Args = {std::shared_ptr<gtsam::DiscreteConditional>}]’
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build/wrap/gtsam/gtsam_wrapper.cpp:14278:97:   required from here
/usr/include/c++/9/ext/new_allocator.h:146:4: error: no matching function for call to ‘gtsam::DiscreteConditional::DiscreteConditional(std::shared_ptr<gtsam::DiscreteConditional>)’
  146 |  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
      |    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteBayesNet.h:21,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build/wrap/gtsam/gtsam_wrapper.cpp:20:
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:121:3: note: candidate: ‘gtsam::DiscreteConditional::DiscreteConditional(const gtsam::DecisionTreeFactor&, const gtsam::DecisionTreeFactor&, const gtsam::Ordering&)’
  121 |   DiscreteConditional(const DecisionTreeFactor& joint,
      |   ^~~~~~~~~~~~~~~~~~~
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:121:3: note:   candidate expects 3 arguments, 1 provided
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:113:3: note: candidate: ‘gtsam::DiscreteConditional::DiscreteConditional(const gtsam::DecisionTreeFactor&, const gtsam::DecisionTreeFactor&)’
  113 |   DiscreteConditional(const DecisionTreeFactor& joint,
      |   ^~~~~~~~~~~~~~~~~~~
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:113:3: note:   candidate expects 2 arguments, 1 provided
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:106:3: note: candidate: ‘gtsam::DiscreteConditional::DiscreteConditional(const DiscreteKey&, const string&)’
  106 |   DiscreteConditional(const DiscreteKey& key, const std::string& spec)
      |   ^~~~~~~~~~~~~~~~~~~
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:106:3: note:   candidate expects 2 arguments, 1 provided
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:101:3: note: candidate: ‘gtsam::DiscreteConditional::DiscreteConditional(const DiscreteKey&, const gtsam::DiscreteKeys&, const string&)’
  101 |   DiscreteConditional(const DiscreteKey& key, const DiscreteKeys& parents,
      |   ^~~~~~~~~~~~~~~~~~~
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:101:3: note:   candidate expects 3 arguments, 1 provided
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:87:3: note: candidate: ‘gtsam::DiscreteConditional::DiscreteConditional(const DiscreteKey&, const gtsam::DiscreteKeys&, const std::vector<double>&)’
   87 |   DiscreteConditional(const DiscreteKey& key, const DiscreteKeys& parents,
      |   ^~~~~~~~~~~~~~~~~~~
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:87:3: note:   candidate expects 3 arguments, 1 provided
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:76:3: note: candidate: ‘gtsam::DiscreteConditional::DiscreteConditional(const DiscreteKey&, const gtsam::DiscreteKeys&, const Table&)’
   76 |   DiscreteConditional(const DiscreteKey& key, const DiscreteKeys& parents,
      |   ^~~~~~~~~~~~~~~~~~~
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:76:3: note:   candidate expects 3 arguments, 1 provided
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:67:12: note: candidate: ‘gtsam::DiscreteConditional::DiscreteConditional(const gtsam::Signature&)’
   67 |   explicit DiscreteConditional(const Signature& signature);
      |            ^~~~~~~~~~~~~~~~~~~
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:67:49: note:   no known conversion for argument 1 from ‘std::shared_ptr<gtsam::DiscreteConditional>’ to ‘const gtsam::Signature&’
   67 |   explicit DiscreteConditional(const Signature& signature);
      |                                ~~~~~~~~~~~~~~~~~^~~~~~~~~
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:63:3: note: candidate: ‘gtsam::DiscreteConditional::DiscreteConditional(size_t, const gtsam::DiscreteKeys&, const ADT&)’
   63 |   DiscreteConditional(size_t nFrontals, const DiscreteKeys& keys,
      |   ^~~~~~~~~~~~~~~~~~~
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:63:3: note:   candidate expects 3 arguments, 1 provided
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:57:3: note: candidate: ‘gtsam::DiscreteConditional::DiscreteConditional(size_t, const gtsam::DiscreteFactor&)’
   57 |   DiscreteConditional(size_t nFrontals, const DiscreteFactor& f);
      |   ^~~~~~~~~~~~~~~~~~~
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:57:3: note:   candidate expects 2 arguments, 1 provided
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:54:3: note: candidate: ‘gtsam::DiscreteConditional::DiscreteConditional()’
   54 |   DiscreteConditional() {}
      |   ^~~~~~~~~~~~~~~~~~~
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:54:3: note:   candidate expects 0 arguments, 1 provided
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:37:20: note: candidate: ‘gtsam::DiscreteConditional::DiscreteConditional(const gtsam::DiscreteConditional&)’
   37 | class GTSAM_EXPORT DiscreteConditional
      |                    ^~~~~~~~~~~~~~~~~~~
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:37:20: note:   no known conversion for argument 1 from ‘std::shared_ptr<gtsam::DiscreteConditional>’ to ‘const gtsam::DiscreteConditional&’
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:37:20: note: candidate: ‘gtsam::DiscreteConditional::DiscreteConditional(gtsam::DiscreteConditional&&)’
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DiscreteConditional.h:37:20: note:   no known conversion for argument 1 from ‘std::shared_ptr<gtsam::DiscreteConditional>’ to ‘gtsam::DiscreteConditional&&’
In file included from /usr/include/x86_64-linux-gnu/c++/9/bits/c++allocator.h:33,
                 from /usr/include/c++/9/bits/allocator.h:46,
                 from /usr/include/c++/9/string:41,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/base/types.h:28,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/global_includes.h:22,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/base/Vector.h:29,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/wrap/matlab.h:25,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build/wrap/gtsam/gtsam_wrapper.cpp:1:
/usr/include/c++/9/ext/new_allocator.h: In instantiation of ‘void __gnu_cxx::new_allocator<_Tp>::construct(_Up*, _Args&& ...) [with _Up = gtsam::DecisionTreeFactor; _Args = {std::shared_ptr<gtsam::DiscreteFactor>}; _Tp = gtsam::DecisionTreeFactor]’:
/usr/include/c++/9/bits/alloc_traits.h:483:4:   required from ‘static void std::allocator_traits<std::allocator<_CharT> >::construct(std::allocator_traits<std::allocator<_CharT> >::allocator_type&, _Up*, _Args&& ...) [with _Up = gtsam::DecisionTreeFactor; _Args = {std::shared_ptr<gtsam::DiscreteFactor>}; _Tp = gtsam::DecisionTreeFactor; std::allocator_traits<std::allocator<_CharT> >::allocator_type = std::allocator<gtsam::DecisionTreeFactor>]’
/usr/include/c++/9/bits/shared_ptr_base.h:548:39:   required from ‘std::_Sp_counted_ptr_inplace<_Tp, _Alloc, _Lp>::_Sp_counted_ptr_inplace(_Alloc, _Args&& ...) [with _Args = {std::shared_ptr<gtsam::DiscreteFactor>}; _Tp = gtsam::DecisionTreeFactor; _Alloc = std::allocator<gtsam::DecisionTreeFactor>; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/9/bits/shared_ptr_base.h:679:16:   required from ‘std::__shared_count<_Lp>::__shared_count(_Tp*&, std::_Sp_alloc_shared_tag<_Alloc>, _Args&& ...) [with _Tp = gtsam::DecisionTreeFactor; _Alloc = std::allocator<gtsam::DecisionTreeFactor>; _Args = {std::shared_ptr<gtsam::DiscreteFactor>}; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/9/bits/shared_ptr_base.h:1344:71:   required from ‘std::__shared_ptr<_Tp, _Lp>::__shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<gtsam::DecisionTreeFactor>; _Args = {std::shared_ptr<gtsam::DiscreteFactor>}; _Tp = gtsam::DecisionTreeFactor; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
/usr/include/c++/9/bits/shared_ptr.h:359:59:   required from ‘std::shared_ptr<_Tp>::shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<gtsam::DecisionTreeFactor>; _Args = {std::shared_ptr<gtsam::DiscreteFactor>}; _Tp = gtsam::DecisionTreeFactor]’
/usr/include/c++/9/bits/shared_ptr.h:701:14:   required from ‘std::shared_ptr<_Tp> std::allocate_shared(const _Alloc&, _Args&& ...) [with _Tp = gtsam::DecisionTreeFactor; _Alloc = std::allocator<gtsam::DecisionTreeFactor>; _Args = {std::shared_ptr<gtsam::DiscreteFactor>}]’
/usr/include/c++/9/bits/shared_ptr.h:717:39:   required from ‘std::shared_ptr<_Tp> std::make_shared(_Args&& ...) [with _Tp = gtsam::DecisionTreeFactor; _Args = {std::shared_ptr<gtsam::DiscreteFactor>}]’
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build/wrap/gtsam/gtsam_wrapper.cpp:14919:86:   required from here
/usr/include/c++/9/ext/new_allocator.h:146:4: error: no matching function for call to ‘gtsam::DecisionTreeFactor::DecisionTreeFactor(std::shared_ptr<gtsam::DiscreteFactor>)’
  146 |  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
      |    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build/wrap/gtsam/gtsam_wrapper.cpp:19:
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DecisionTreeFactor.h:120:14: note: candidate: ‘gtsam::DecisionTreeFactor::DecisionTreeFactor(const gtsam::DiscreteConditional&)’
  120 |     explicit DecisionTreeFactor(const DiscreteConditional& c);
      |              ^~~~~~~~~~~~~~~~~~
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DecisionTreeFactor.h:120:60: note:   no known conversion for argument 1 from ‘std::shared_ptr<gtsam::DiscreteFactor>’ to ‘const gtsam::DiscreteConditional&’
  120 |     explicit DecisionTreeFactor(const DiscreteConditional& c);
      |                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~^
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DecisionTreeFactor.h:116:5: note: candidate: ‘gtsam::DecisionTreeFactor::DecisionTreeFactor(const DiscreteKey&, const std::vector<double>&)’
  116 |     DecisionTreeFactor(const DiscreteKey& key, const std::vector<double>& row)
      |     ^~~~~~~~~~~~~~~~~~
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DecisionTreeFactor.h:116:5: note:   candidate expects 2 arguments, 1 provided
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DecisionTreeFactor.h:112:5: note: candidate: ‘template<class SOURCE> gtsam::DecisionTreeFactor::DecisionTreeFactor(const DiscreteKey&, SOURCE)’
  112 |     DecisionTreeFactor(const DiscreteKey& key, SOURCE table)
      |     ^~~~~~~~~~~~~~~~~~
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DecisionTreeFactor.h:112:5: note:   template argument deduction/substitution failed:
In file included from /usr/include/x86_64-linux-gnu/c++/9/bits/c++allocator.h:33,
                 from /usr/include/c++/9/bits/allocator.h:46,
                 from /usr/include/c++/9/string:41,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/base/types.h:28,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/global_includes.h:22,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/base/Vector.h:29,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/wrap/matlab.h:25,
                 from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build/wrap/gtsam/gtsam_wrapper.cpp:1:
/usr/include/c++/9/ext/new_allocator.h:146:4: note:   candidate expects 2 arguments, 1 provided
  146 |  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
      |    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build/wrap/gtsam/gtsam_wrapper.cpp:19:
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DecisionTreeFactor.h:108:5: note: candidate: ‘gtsam::DecisionTreeFactor::DecisionTreeFactor(const gtsam::DiscreteKeys&, const string&)’
  108 |     DecisionTreeFactor(const DiscreteKeys& keys, const std::string& table);
      |     ^~~~~~~~~~~~~~~~~~
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DecisionTreeFactor.h:108:5: note:   candidate expects 2 arguments, 1 provided
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DecisionTreeFactor.h:87:5: note: candidate: ‘gtsam::DecisionTreeFactor::DecisionTreeFactor(const gtsam::DiscreteKeys&, const std::vector<double>&)’
   87 |     DecisionTreeFactor(const DiscreteKeys& keys,
      |     ^~~~~~~~~~~~~~~~~~
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DecisionTreeFactor.h:87:5: note:   candidate expects 2 arguments, 1 provided
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DecisionTreeFactor.h:66:5: note: candidate: ‘gtsam::DecisionTreeFactor::DecisionTreeFactor(const gtsam::DiscreteKeys&, const ADT&)’
   66 |     DecisionTreeFactor(const DiscreteKeys& keys, const ADT& potentials);
      |     ^~~~~~~~~~~~~~~~~~
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DecisionTreeFactor.h:66:5: note:   candidate expects 2 arguments, 1 provided
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DecisionTreeFactor.h:63:5: note: candidate: ‘gtsam::DecisionTreeFactor::DecisionTreeFactor()’
   63 |     DecisionTreeFactor();
      |     ^~~~~~~~~~~~~~~~~~
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DecisionTreeFactor.h:63:5: note:   candidate expects 0 arguments, 1 provided
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DecisionTreeFactor.h:45:22: note: candidate: ‘gtsam::DecisionTreeFactor::DecisionTreeFactor(const gtsam::DecisionTreeFactor&)’
   45 |   class GTSAM_EXPORT DecisionTreeFactor : public DiscreteFactor,
      |                      ^~~~~~~~~~~~~~~~~~
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DecisionTreeFactor.h:45:22: note:   no known conversion for argument 1 from ‘std::shared_ptr<gtsam::DiscreteFactor>’ to ‘const gtsam::DecisionTreeFactor&’
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DecisionTreeFactor.h:45:22: note: candidate: ‘gtsam::DecisionTreeFactor::DecisionTreeFactor(gtsam::DecisionTreeFactor&&)’
/Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam/discrete/DecisionTreeFactor.h:45:22: note:   no known conversion for argument 1 from ‘std::shared_ptr<gtsam::DiscreteFactor>’ to ‘gtsam::DecisionTreeFactor&&’
make[2]: *** [matlab/CMakeFiles/gtsam_matlab_wrapper.dir/build.make:96: matlab/CMakeFiles/gtsam_matlab_wrapper.dir/__/wrap/gtsam/gtsam_wrapper.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:34652: matlab/CMakeFiles/gtsam_matlab_wrapper.dir/all] Error 2
make: *** [Makefile:166: all] Error 2
```

## Environment

Using a conda environment
gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Python 3.12.4
R2024b Update 1 (24.2.0.2740171) 64-bit (glnxa64)
cmake version 3.30.5
boost_1.87.0

## Adicional Information
Build results:
```
-- GTSAM_POSE3_EXPMAP=ON, enabling GTSAM_ROT3_EXPMAP as well
-- Found Eigen version: 3.4.0
CMake Deprecation Warning at gtsam/3rdparty/metis/CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- checking for thread-local storage - found
-- Could NOT find MKL (missing: MKL_INCLUDE_DIR MKL_LIBRARIES) 
-- Building 3rdparty
-- Could NOT find GeographicLib (missing: GeographicLib_LIBRARY_DIRS GeographicLib_LIBRARIES GeographicLib_INCLUDE_DIRS) 
-- Building base
-- Building basis
-- Building constrained
-- Building geometry
-- Building inference
-- Building symbolic
-- Building discrete
-- Building hybrid
-- Building linear
-- Building nonlinear
-- Building sam
-- Building sfm
-- Building slam
-- Building navigation
-- GTSAM Version: 4.3a0
-- Install prefix: /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam_install
-- Building GTSAM - as a SHARED library
-- Building base_unstable
-- Building geometry_unstable
-- Building linear_unstable
-- Building discrete_unstable
-- Building dynamics_unstable
-- Building nonlinear_unstable
-- Building slam_unstable
-- Building partition_unstable
-- GTSAM_UNSTABLE Version: 4.3a0
-- Install prefix: /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam_install
-- gtwrap Package config : /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam_install/lib/cmake/gtwrap
-- gtwrap version        : 1.0
-- gtwrap CMake path     : /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam_install/lib/cmake/gtwrap
-- gtwrap library path   : /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam_install/lib/gtwrap
-- gtwrap binary path    : /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam_install/bin/gtwrap
-- gtwrap header path    : /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam_install/include/gtwrap
-- Building wrap module gtsam
CMake Warning at wrap/cmake/MatlabWrap.cmake:225 (find_package):
  By not providing "FindPythonInterp.cmake" in CMAKE_MODULE_PATH this project
  has asked CMake to find a package configuration file provided by
  "PythonInterp", but CMake did not find one.

  Could not find a package configuration file provided by "PythonInterp"
  (requested version 3.12.4) with any of the following names:

    PythonInterpConfig.cmake
    pythoninterp-config.cmake

  Add the installation prefix of "PythonInterp" to CMAKE_PREFIX_PATH or set
  "PythonInterp_DIR" to a directory containing one of the above files.  If
  "PythonInterp" provides a separate development package or SDK, be sure it
  has been installed.
Call Stack (most recent call first):
  wrap/cmake/MatlabWrap.cmake:92 (wrap_library_internal)
  wrap/cmake/MatlabWrap.cmake:68 (wrap_and_install_library)
  matlab/CMakeLists.txt:89 (matlab_wrap)


CMake Warning at wrap/cmake/MatlabWrap.cmake:226 (find_package):
  By not providing "FindPythonLibs.cmake" in CMAKE_MODULE_PATH this project
  has asked CMake to find a package configuration file provided by
  "PythonLibs", but CMake did not find one.

  Could not find a package configuration file provided by "PythonLibs"
  (requested version 3.12.4) with any of the following names:

    PythonLibsConfig.cmake
    pythonlibs-config.cmake

  Add the installation prefix of "PythonLibs" to CMAKE_PREFIX_PATH or set
  "PythonLibs_DIR" to a directory containing one of the above files.  If
  "PythonLibs" provides a separate development package or SDK, be sure it has
  been installed.
Call Stack (most recent call first):
  wrap/cmake/MatlabWrap.cmake:92 (wrap_library_internal)
  wrap/cmake/MatlabWrap.cmake:68 (wrap_and_install_library)
  matlab/CMakeLists.txt:89 (matlab_wrap)


-- Installing Matlab Toolbox to /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/gtsam_install/gtsam_toolbox
-- Wrote /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build/GTSAMConfig.cmake
-- Wrote /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build/GTSAM_UNSTABLEConfig.cmake
-- ===============================================================
-- ================  Configuration Options  ======================
--  CMAKE_CXX_COMPILER_ID type                       : GNU
--  CMAKE_CXX_COMPILER_VERSION                       : 9.4.0
--  CMake version                                    : 3.30.5
--  CMake generator                                  : Unix Makefiles
--  CMake build tool                                 : /usr/bin/make
-- Build flags                                               
--  Build Tests                                      : Enabled
--  Build examples with 'make all'                   : Enabled
--  Build timing scripts with 'make all'             : Disabled
--  Build Docs                                       : Enabled
--  Build shared GTSAM libraries                     : Enabled
--  Put build type in library name                   : Enabled
--  Build libgtsam_unstable                          : Enabled
--  Build GTSAM unstable Python                      : Enabled
--  Build MATLAB Toolbox for unstable                : Disabled
--  Build for native architecture                    : Disabled
--  Build type                                       : Release
--  C compilation flags                              :  -O3 -DNDEBUG
--  C++ compilation flags                            :  -O3 -DNDEBUG
--  Enable Boost serialization                       : ON
--  GTSAM_COMPILE_FEATURES_PUBLIC                    : cxx_std_17
--  GTSAM_COMPILE_OPTIONS_PUBLIC                     : 
--  GTSAM_COMPILE_DEFINITIONS_PUBLIC                 : 
--  GTSAM_COMPILE_OPTIONS_PUBLIC_RELEASE             : 
--  GTSAM_COMPILE_DEFINITIONS_PUBLIC_RELEASE         : 
--  Use System Eigen                                 : OFF (Using version: 3.4.0)
--  Use System Metis                                 : OFF
--  Using Boost version                              : 1.87.0
--  Use Intel TBB                                    : Yes (Version: 2020.1)
--  Eigen will use MKL                               : MKL not found
--  Eigen will use MKL and OpenMP                    : OpenMP found but GTSAM_WITH_EIGEN_MKL is disabled
--  Default allocator                                : TBB
--  Cheirality exceptions enabled                    : YES
--  Build with ccache                                : No
-- Packaging flags
--  CPack Source Generator                           : TGZ
--  CPack Generator                                  : TGZ
-- GTSAM flags                                               
--  Quaternions as default Rot3                      : Disabled
--  Runtime consistency checking                     : Disabled
--  Build with Memory Sanitizer                      : Disabled
--  Rot3 retract is full ExpMap                      : Enabled
--  Pose3 retract is full ExpMap                     : Enabled
--  Enable branch merging in DecisionTree            : Enabled
--  Enable timing machinery                          : Disabled
--  Allow features deprecated in GTSAM 4.3           : Enabled
--  Metis-based Nested Dissection                    : Enabled
--  Use tangent-space preintegration                 : Enabled
-- MATLAB toolbox flags
--  Install MATLAB toolbox                           : Enabled
--  MATLAB root                                      : /Data/miguelcruz/matlab
--  MEX binary                                       : /Data/miguelcruz/matlab/bin/glnxa64/mex
-- Python toolbox flags                                      
--  Build Python module with pybind                  : Disabled
-- ===============================================================
-- Configuring done (0.7s)
-- Generating done (1.5s)
-- Build files have been written to: /Data/miguelcruz/SingleImageDepthEstimation/prepareDS_registration/gtsam/gtsam/build
```

--- a/gtsam/discrete/DiscreteValues.cpp
+++ b/gtsam/discrete/DiscreteValues.cpp
@@ -145,6 +145,11 @@ string DiscreteValues::html(const KeyFormatter& keyFormatter,
 }
 
 /* ************************************************************************ */
+void PrintDiscreteValues(const DiscreteValues& values, const std::string& s,
+                         const KeyFormatter& keyFormatter) {
+  values.print(s, keyFormatter);
+}
+
 string markdown(const DiscreteValues& values, const KeyFormatter& keyFormatter,
                 const DiscreteValues::Names& names) {
   return values.markdown(keyFormatter, names);

--- a/gtsam/discrete/DiscreteValues.h
+++ b/gtsam/discrete/DiscreteValues.h
@@ -188,6 +188,11 @@ inline std::vector<DiscreteValues> cartesianProduct(const DiscreteKeys& keys) {
   return DiscreteValues::CartesianProduct(keys);
 }
 
+/// Free version of print for wrapper
+void PrintDiscreteValues(
+    const DiscreteValues& values, const std::string& s = "",
+    const KeyFormatter& keyFormatter = DefaultKeyFormatter);
+
 /// Free version of markdown.
 std::string GTSAM_EXPORT
 markdown(const DiscreteValues& values,

--- a/gtsam/discrete/DiscreteValues.h
+++ b/gtsam/discrete/DiscreteValues.h
@@ -189,9 +189,9 @@ inline std::vector<DiscreteValues> cartesianProduct(const DiscreteKeys& keys) {
 }
 
 /// Free version of print for wrapper
-void PrintDiscreteValues(
-    const DiscreteValues& values, const std::string& s = "",
-    const KeyFormatter& keyFormatter = DefaultKeyFormatter);
+void GTSAM_EXPORT
+PrintDiscreteValues(const DiscreteValues& values, const std::string& s = "",
+                    const KeyFormatter& keyFormatter = DefaultKeyFormatter);
 
 /// Free version of markdown.
 std::string GTSAM_EXPORT

--- a/gtsam/discrete/discrete.i
+++ b/gtsam/discrete/discrete.i
@@ -23,9 +23,9 @@ class DiscreteKeys {
 std::vector<gtsam::DiscreteValues> cartesianProduct(
     const gtsam::DiscreteKeys& keys);
 
-// void PrintDiscreteValues(
-//     const gtsam::DiscreteValues& values, const std::string& s = "",
-//     const gtsam::KeyFormatter& keyFormatter = gtsam::DefaultKeyFormatter);
+void PrintDiscreteValues(
+    const gtsam::DiscreteValues& values, const std::string& s = "",
+    const gtsam::KeyFormatter& keyFormatter = gtsam::DefaultKeyFormatter);
 
 string markdown(
     const gtsam::DiscreteValues& values,

--- a/gtsam/discrete/discrete.i
+++ b/gtsam/discrete/discrete.i
@@ -477,9 +477,9 @@ class DiscreteSearchSolution {
 };
 
 class DiscreteSearch {
-  static DiscreteSearch FromFactorGraph(const gtsam::DiscreteFactorGraph& factorGraph,
-                                        const gtsam::Ordering& ordering,
-                                        bool buildJunctionTree = false);
+  static gtsam::DiscreteSearch FromFactorGraph(
+      const gtsam::DiscreteFactorGraph& factorGraph,
+      const gtsam::Ordering& ordering, bool buildJunctionTree = false);
 
   DiscreteSearch(const gtsam::DiscreteEliminationTree& etree);
   DiscreteSearch(const gtsam::DiscreteJunctionTree& junctionTree);

--- a/gtsam/discrete/discrete.i
+++ b/gtsam/discrete/discrete.i
@@ -23,9 +23,9 @@ class DiscreteKeys {
 std::vector<gtsam::DiscreteValues> cartesianProduct(
     const gtsam::DiscreteKeys& keys);
 
-void PrintDiscreteValues(
-    const gtsam::DiscreteValues& values, const std::string& s = "",
-    const gtsam::KeyFormatter& keyFormatter = gtsam::DefaultKeyFormatter);
+// void PrintDiscreteValues(
+//     const gtsam::DiscreteValues& values, const std::string& s = "",
+//     const gtsam::KeyFormatter& keyFormatter = gtsam::DefaultKeyFormatter);
 
 string markdown(
     const gtsam::DiscreteValues& values,

--- a/gtsam/discrete/discrete.i
+++ b/gtsam/discrete/discrete.i
@@ -22,6 +22,11 @@ class DiscreteKeys {
 // DiscreteValues is added in specializations/discrete.h as a std::map
 std::vector<gtsam::DiscreteValues> cartesianProduct(
     const gtsam::DiscreteKeys& keys);
+
+void PrintDiscreteValues(
+    const gtsam::DiscreteValues& values, const std::string& s = "",
+    const gtsam::KeyFormatter& keyFormatter = gtsam::DefaultKeyFormatter);
+
 string markdown(
     const gtsam::DiscreteValues& values,
     const gtsam::KeyFormatter& keyFormatter = gtsam::DefaultKeyFormatter);

--- a/gtsam/hybrid/hybrid.i
+++ b/gtsam/hybrid/hybrid.i
@@ -155,6 +155,8 @@ class HybridBayesNet {
   gtsam::GaussianBayesNet choose(const gtsam::DiscreteValues& assignment) const;
 
   gtsam::HybridValues optimize() const;
+  gtsam::VectorValues optimize(const gtsam::DiscreteValues& assignment) const;
+
   gtsam::HybridValues sample(const gtsam::HybridValues& given) const;
   gtsam::HybridValues sample() const;
 

--- a/gtsam/hybrid/hybrid.i
+++ b/gtsam/hybrid/hybrid.i
@@ -152,6 +152,8 @@ class HybridBayesNet {
   gtsam::HybridGaussianFactorGraph toFactorGraph(
       const gtsam::VectorValues& measurements) const;
 
+  gtsam::GaussianBayesNet choose(const gtsam::DiscreteValues& assignment) const;
+
   gtsam::HybridValues optimize() const;
   gtsam::HybridValues sample(const gtsam::HybridValues& given) const;
   gtsam::HybridValues sample() const;

--- a/python/gtsam/examples/HybridCity10000.py
+++ b/python/gtsam/examples/HybridCity10000.py
@@ -334,6 +334,17 @@ class Experiment:
 
         for idx, assignment in enumerate(all_assignments):
             result = gtsam.Values()
+            gbn = self.smoother_.hybridBayesNet().choose(assignment)
+
+            # Check to see if the GBN has any nullptrs, if it does it is null overall
+            is_invalid_gbn = False
+            for i in range(gbn.size()):
+                if gbn.at(i) is None:
+                    is_invalid_gbn = True
+                    break
+            if is_invalid_gbn:
+                continue
+
             delta = self.smoother_.hybridBayesNet().optimize(assignment)
             result.insert_or_assign(self.initial_.retract(delta))
 

--- a/python/gtsam/examples/HybridCity10000.py
+++ b/python/gtsam/examples/HybridCity10000.py
@@ -64,7 +64,7 @@ prior_noise_model = gtsam.noiseModel.Diagonal.Sigmas(
     np.asarray([0.0001, 0.0001, 0.0001]))
 
 pose_noise_model = gtsam.noiseModel.Diagonal.Sigmas(
-    np.asarray([1.0 / 30.0, 1.0 / 30.0, 1.0 / 100.0]))
+    np.asarray([1.0 / 20.0, 1.0 / 20.0, 1.0 / 100.0]))
 pose_noise_constant = pose_noise_model.negLogConstant()
 
 

--- a/python/gtsam/examples/HybridCity10000.py
+++ b/python/gtsam/examples/HybridCity10000.py
@@ -117,24 +117,33 @@ class City10000Dataset:
 
 def plot_all_results(ground_truth,
                      all_results,
+                     iters=0,
                      estimate_color=(0.1, 0.1, 0.9, 0.4),
                      estimate_label="Hybrid Factor Graphs",
-                     text=""):
+                     text="",
+                     filename="city10000_results.svg"):
     """Plot the City10000 estimates against the ground truth.
 
     Args:
         ground_truth: The ground truth trajectory as xy values.
-        all_results (List[Tuple(np.ndarray, str)]): All the estimates trajectory as xy values, as well as assginment strings.
+        all_results (List[Tuple(np.ndarray, str)]): All the estimates trajectory as xy values,
+            as well as assginment strings.
         estimate_color (tuple, optional): The color to use for the graph of estimates.
             Defaults to (0.1, 0.1, 0.9, 0.4).
         estimate_label (str, optional): Label for the estimates, used in the legend.
             Defaults to "Hybrid Factor Graphs".
     """
-    fig, axes = plt.subplots(int(np.ceil(len(all_results) / 2)), 2)
+    if len(all_results) == 1:
+        fig, axes = plt.subplots(1, 1)
+        axes = [axes]
+    else:
+        fig, axes = plt.subplots(int(np.ceil(len(all_results) / 2)), 2)
+        axes = axes.flatten()
+
     for i, (estimates, s) in enumerate(all_results):
         ax = axes[i]
         ax.axis('equal')
-        ax.axis((-75.0, 75.0, -75.0, 75.0))
+        ax.axis((-75.0, 100.0, -75.0, 75.0))
 
         gt = ground_truth[:estimates.shape[0]]
         ax.plot(gt[:, 0],
@@ -149,14 +158,17 @@ def plot_all_results(ground_truth,
                 linewidth=1,
                 color=estimate_color,
                 label=estimate_label)
-        ax.legend()
-        ax.text(0.0, 100.0, s)
+        # ax.legend()
+        # Plot text `s` at (x, y) on axis
+        ax.text(-60.0, 60.0, s)
+
+    fig.suptitle(f"After {iters} iterations")
 
     num_chunks = int(np.ceil(len(text) / 90))
     text = "\n".join(text[i * 60:(i + 1) * 60] for i in range(num_chunks))
     fig.text(0.0, 0.015, s=text)
 
-    fig.savefig("city10000_results.svg", format="svg")
+    fig.savefig(filename, format="svg")
 
 
 class Experiment:
@@ -362,9 +374,9 @@ class Experiment:
                 # TODO Get cardinality from DiscreteFactor
                 discrete_keys.push_back((key, 2))
             print("plotting all hypotheses")
-            self.plot_all_hypotheses(discrete_keys, key_t + 1)
+            self.plot_all_hypotheses(discrete_keys, key_t + 1, index)
 
-    def plot_all_hypotheses(self, discrete_keys, num_poses):
+    def plot_all_hypotheses(self, discrete_keys, num_poses, num_iters=0):
         """Plot all possible hypotheses."""
 
         # Get ground truth
@@ -410,7 +422,11 @@ class Experiment:
             ])
             all_results.append((poses, assignment_string))
 
-        plot_all_results(gt, all_results, text=fixed_values_str)
+        plot_all_results(gt,
+                         all_results,
+                         iters=num_iters,
+                         text=fixed_values_str,
+                         filename=f"city10000_results_{num_iters}.svg")
 
     def save_results(self, result, final_key, time_list):
         """Save results to file."""

--- a/python/gtsam/examples/HybridCity10000.py
+++ b/python/gtsam/examples/HybridCity10000.py
@@ -47,6 +47,11 @@ def parse_arguments():
         type=int,
         default=10,
         help="The maximum number of hypotheses to keep at any time.")
+    parser.add_argument(
+        "-p",
+        action="store_true",
+        help="Plot all hypotheses. NOTE: This is exponential, use with caution."
+    )
     return parser.parse_args()
 
 
@@ -125,7 +130,9 @@ def plot_estimates(gt,
     fig = plt.figure(fignum)
     ax = fig.gca()
     ax.axis('equal')
-    ax.axis((-65.0, 65.0, -75.0, 60.0))
+    ax.axis((-75.0, 75.0, -75.0, 75.0))
+
+    gt = gt[:estimates.shape[0]]
     ax.plot(gt[:, 0],
             gt[:, 1],
             '--',
@@ -150,11 +157,12 @@ class Experiment:
 
     def __init__(self,
                  filename: str,
-                 marginal_threshold: float = 1.9999,
+                 marginal_threshold: float = 0.9999,
                  max_loop_count: int = 150,
                  update_frequency: int = 3,
                  max_num_hypotheses: int = 10,
-                 relinearization_frequency: int = 10):
+                 relinearization_frequency: int = 10,
+                 plot_hypotheses: bool = False):
         self.dataset_ = City10000Dataset(filename)
         self.max_loop_count = max_loop_count
         self.update_frequency = update_frequency
@@ -165,6 +173,8 @@ class Experiment:
         self.new_factors_ = HybridNonlinearFactorGraph()
         self.all_factors_ = HybridNonlinearFactorGraph()
         self.initial_ = Values()
+        
+        self.plot_hypotheses: = plot_hypotheses
 
     def hybrid_loop_closure_factor(self, loop_counter, key_s, key_t,
                                    measurement: Pose2):
@@ -215,7 +225,7 @@ class Experiment:
         bayesNet = linearized.eliminateSequential()
         delta: HybridValues = bayesNet.optimize()
         self.initial_ = self.initial_.retract(delta.continuous())
-        self.smoother_.reinitialize(bayesNet)
+        self.smoother_.reInitialize(bayesNet)
         after_update = time.time()
         print(f"Took {after_update - before_update} seconds.")
         return after_update - before_update
@@ -331,13 +341,14 @@ class Experiment:
 
         # self.save_results(result, key_t + 1, time_list)
 
-        # Get all the discrete values
-        discrete_keys = gtsam.DiscreteKeys()
-        for key in delta.discrete().keys():
-            # TODO Get cardinality from DiscreteFactor
-            discrete_keys.push_back((key, 2))
-        print("plotting all hypotheses")
-        self.plot_all_hypotheses(discrete_keys, key_t + 1)
+        if self.plot_hypotheses:
+            # Get all the discrete values
+            discrete_keys = gtsam.DiscreteKeys()
+            for key in delta.discrete().keys():
+                # TODO Get cardinality from DiscreteFactor
+                discrete_keys.push_back((key, 2))
+            print("plotting all hypotheses")
+            self.plot_all_hypotheses(discrete_keys, key_t + 1)
 
     def plot_all_hypotheses(self, discrete_keys, num_poses):
         """Plot all possible hypotheses."""
@@ -346,7 +357,6 @@ class Experiment:
         gt = np.loadtxt(gtsam.findExampleDataFile("ISAM2_GT_city10000.txt"),
                         delimiter=" ")
 
-        # print(discrete_keys)
         # Get all possible assignments
         all_assignments = gtsam.cartesianProduct(discrete_keys)
 
@@ -428,7 +438,8 @@ def main():
     experiment = Experiment(gtsam.findExampleDataFile(args.data_file),
                             max_loop_count=args.max_loop_count,
                             update_frequency=args.update_frequency,
-                            max_num_hypotheses=args.max_num_hypotheses)
+                            max_num_hypotheses=args.max_num_hypotheses,
+                            plot_hypotheses=args.plot_hypotheses)
     experiment.run()
 
 

--- a/python/gtsam/examples/HybridCity10000.py
+++ b/python/gtsam/examples/HybridCity10000.py
@@ -147,7 +147,7 @@ class Experiment:
         after_update = time.time()
         return after_update - before_update
 
-    def reInitialize(self) -> float:
+    def reinitialize(self) -> float:
         """Re-linearize, solve ALL, and re-initialize smoother."""
         print(f"================= Re-Initialize: {self.all_factors_.size()}")
         before_update = time.time()
@@ -157,7 +157,7 @@ class Experiment:
         bayesNet = linearized.eliminateSequential()
         delta: HybridValues = bayesNet.optimize()
         self.initial_ = self.initial_.retract(delta.continuous())
-        self.smoother_.reInitialize(bayesNet)
+        self.smoother_.reinitialize(bayesNet)
         after_update = time.time()
         print(f"Took {after_update - before_update} seconds.")
         return after_update - before_update
@@ -240,7 +240,7 @@ class Experiment:
                 update_count += 1
 
                 if update_count % self.relinearization_frequency == 0:
-                    self.reInitialize()
+                    self.reinitialize()
 
             #  Record timing for odometry edges only
             if key_s == key_t - 1:
@@ -271,8 +271,12 @@ class Experiment:
         total_time = end_time - start_time
         print(f"Total time: {total_time} seconds")
 
+        self.save_results(result, key_t + 1, time_list)
+
+    def save_results(self, result, final_key, time_list):
+        """Save results to file."""
         # Write results to file
-        self.write_result(result, key_t + 1, "Hybrid_City10000.txt")
+        self.write_result(result, final_key, "Hybrid_City10000.txt")
 
         # Write timing info to file
         self.write_timing_info(time_list=time_list)

--- a/python/gtsam/examples/HybridCity10000.py
+++ b/python/gtsam/examples/HybridCity10000.py
@@ -112,7 +112,7 @@ class City10000Dataset:
         if line:
             return self.parse_line(line)
         else:
-            return None, None
+            return None, None, None
 
 
 def plot_all_results(ground_truth,
@@ -140,7 +140,7 @@ def plot_all_results(ground_truth,
         fig, axes = plt.subplots(int(np.ceil(len(all_results) / 2)), 2)
         axes = axes.flatten()
 
-    for i, (estimates, s) in enumerate(all_results):
+    for i, (estimates, s, prob) in enumerate(all_results):
         ax = axes[i]
         ax.axis('equal')
         ax.axis((-75.0, 100.0, -75.0, 75.0))
@@ -159,14 +159,18 @@ def plot_all_results(ground_truth,
                 color=estimate_color,
                 label=estimate_label)
         # ax.legend()
-        # Plot text `s` at (x, y) on axis
-        ax.text(-60.0, 60.0, s)
+        ax.set_title(f"P={prob:.3f}\n{s}", fontdict={'fontsize': 10})
 
     fig.suptitle(f"After {iters} iterations")
 
     num_chunks = int(np.ceil(len(text) / 90))
     text = "\n".join(text[i * 60:(i + 1) * 60] for i in range(num_chunks))
-    fig.text(0.0, 0.015, s=text)
+    fig.text(0.5,
+             0.015,
+             s=text,
+             wrap=True,
+             horizontalalignment='center',
+             fontsize=12)
 
     fig.savefig(filename, format="svg")
 
@@ -388,7 +392,7 @@ class Experiment:
             key, cardinality = discrete_keys.at(i)
             if key not in self.smoother_.fixedValues().keys():
                 dkeys.push_back((key, cardinality))
-        fixed_values_str = "_".join(
+        fixed_values_str = " ".join(
             f"{gtsam.DefaultKeyFormatter(k)}:{v}"
             for k, v in self.smoother_.fixedValues().items())
 
@@ -420,7 +424,19 @@ class Experiment:
                 f"{gtsam.DefaultKeyFormatter(k)}={v}"
                 for k, v in assignment.items()
             ])
-            all_results.append((poses, assignment_string))
+
+            conditional = self.smoother_.hybridBayesNet().at(
+                self.smoother_.hybridBayesNet().size() - 1).asDiscrete()
+            discrete_values = self.smoother_.fixedValues()
+            for k, v in assignment.items():
+                discrete_values[k] = v
+
+            if conditional is None:
+                probability = 1.0
+            else:
+                probability = conditional.evaluate(discrete_values)
+
+            all_results.append((poses, assignment_string, probability))
 
         plot_all_results(gt,
                          all_results,

--- a/python/gtsam/examples/HybridCity10000.py
+++ b/python/gtsam/examples/HybridCity10000.py
@@ -48,6 +48,7 @@ def parse_arguments():
         default=10,
         help="The maximum number of hypotheses to keep at any time.")
     parser.add_argument(
+        "--plot_hypotheses",
         "-p",
         action="store_true",
         help="Plot all hypotheses. NOTE: This is exponential, use with caution."
@@ -173,8 +174,8 @@ class Experiment:
         self.new_factors_ = HybridNonlinearFactorGraph()
         self.all_factors_ = HybridNonlinearFactorGraph()
         self.initial_ = Values()
-        
-        self.plot_hypotheses: = plot_hypotheses
+
+        self.plot_hypotheses = plot_hypotheses
 
     def hybrid_loop_closure_factor(self, loop_counter, key_s, key_t,
                                    measurement: Pose2):
@@ -358,6 +359,10 @@ class Experiment:
                         delimiter=" ")
 
         # Get all possible assignments
+        if discrete_keys.size() > 5:
+            print("Too many discrete keys to plot all hypotheses. Exiting.")
+            exit(0)
+
         all_assignments = gtsam.cartesianProduct(discrete_keys)
 
         for idx, assignment in enumerate(all_assignments):

--- a/python/gtsam/examples/HybridCity10000.py
+++ b/python/gtsam/examples/HybridCity10000.py
@@ -85,12 +85,15 @@ class City10000Dataset:
         """Read a `line` from the dataset, separated by the `delimiter`."""
         return line.split(delimiter)
 
-    def parse_line(self, line: str) -> tuple[list[Pose2], tuple[int, int]]:
+    def parse_line(self,
+                   line: str) -> tuple[list[Pose2], tuple[int, int], bool]:
         """Parse line from file"""
         parts = self.read_line(line)
 
         key_s = int(parts[1])
         key_t = int(parts[3])
+
+        is_ambiguous_loop = bool(int(parts[4]))
 
         num_measurements = int(parts[5])
         pose_array = [Pose2()] * num_measurements
@@ -101,7 +104,7 @@ class City10000Dataset:
             rad = float(parts[8 + 3 * i])
             pose_array[i] = Pose2(x, y, rad)
 
-        return pose_array, (key_s, key_t)
+        return pose_array, (key_s, key_t), is_ambiguous_loop
 
     def next(self):
         """Read and parse the next line."""
@@ -126,7 +129,6 @@ def plot_all_results(ground_truth,
         estimate_label (str, optional): Label for the estimates, used in the legend.
             Defaults to "Hybrid Factor Graphs".
     """
-    print(len(all_results))
     fig, axes = plt.subplots(int(np.ceil(len(all_results) / 2)), 2)
     for i, (estimates, text) in enumerate(all_results):
         ax = axes[i]
@@ -259,7 +261,7 @@ class Experiment:
         start_time = time.time()
 
         while index < self.max_loop_count:
-            pose_array, keys = self.dataset_.next()
+            pose_array, keys, is_ambiguous_loop = self.dataset_.next()
             if pose_array is None:
                 break
             key_s = keys[0]
@@ -293,8 +295,14 @@ class Experiment:
                     self.initial_.atPose2(X(key_s)) * odom_pose)
             else:
                 # Loop closure
-                loop_factor = self.hybrid_loop_closure_factor(
-                    loop_count, key_s, key_t, odom_pose)
+                if is_ambiguous_loop:
+                    loop_factor = self.hybrid_loop_closure_factor(
+                        loop_count, key_s, key_t, odom_pose)
+
+                else:
+                    loop_factor = BetweenFactorPose2(X(key_s), X(key_t),
+                                                     odom_pose,
+                                                     pose_noise_model)
 
                 # print loop closure event keys:
                 print(f"Loop closure: {key_s} {key_t}")

--- a/python/gtsam/examples/HybridCity10000.py
+++ b/python/gtsam/examples/HybridCity10000.py
@@ -118,7 +118,8 @@ class City10000Dataset:
 def plot_all_results(ground_truth,
                      all_results,
                      estimate_color=(0.1, 0.1, 0.9, 0.4),
-                     estimate_label="Hybrid Factor Graphs"):
+                     estimate_label="Hybrid Factor Graphs",
+                     text=""):
     """Plot the City10000 estimates against the ground truth.
 
     Args:
@@ -130,7 +131,7 @@ def plot_all_results(ground_truth,
             Defaults to "Hybrid Factor Graphs".
     """
     fig, axes = plt.subplots(int(np.ceil(len(all_results) / 2)), 2)
-    for i, (estimates, text) in enumerate(all_results):
+    for i, (estimates, s) in enumerate(all_results):
         ax = axes[i]
         ax.axis('equal')
         ax.axis((-75.0, 75.0, -75.0, 75.0))
@@ -149,7 +150,11 @@ def plot_all_results(ground_truth,
                 color=estimate_color,
                 label=estimate_label)
         ax.legend()
-        ax.text(0.0, 100.0, text)
+        ax.text(0.0, 100.0, s)
+
+    num_chunks = int(np.ceil(len(text) / 90))
+    text = "\n".join(text[i * 60:(i + 1) * 60] for i in range(num_chunks))
+    fig.text(0.0, 0.015, s=text)
 
     fig.savefig("city10000_results.svg", format="svg")
 
@@ -371,6 +376,9 @@ class Experiment:
             key, cardinality = discrete_keys.at(i)
             if key not in self.smoother_.fixedValues().keys():
                 dkeys.push_back((key, cardinality))
+        fixed_values_str = "_".join(
+            f"{gtsam.DefaultKeyFormatter(k)}:{v}"
+            for k, v in self.smoother_.fixedValues().items())
 
         all_assignments = gtsam.cartesianProduct(dkeys)
 
@@ -402,7 +410,7 @@ class Experiment:
             ])
             all_results.append((poses, assignment_string))
 
-        plot_all_results(gt, all_results)
+        plot_all_results(gt, all_results, text=fixed_values_str)
 
     def save_results(self, result, final_key, time_list):
         """Save results to file."""

--- a/python/gtsam/examples/HybridCity10000.py
+++ b/python/gtsam/examples/HybridCity10000.py
@@ -29,6 +29,24 @@ def parse_arguments():
     parser.add_argument("--data_file",
                         help="The path to the City10000 data file",
                         default="T1_city10000_04.txt")
+    parser.add_argument(
+        "--max_loop_count",
+        "-l",
+        type=int,
+        default=10000,
+        help="The maximum number of loops to run over the dataset")
+    parser.add_argument(
+        "--update_frequency",
+        "-u",
+        type=int,
+        default=3,
+        help="After how many steps to run the smoother update.")
+    parser.add_argument(
+        "--max_num_hypotheses",
+        "-m",
+        type=int,
+        default=10,
+        help="The maximum number of hypotheses to keep at any time.")
     return parser.parse_args()
 
 
@@ -407,7 +425,10 @@ def main():
     """Main runner"""
     args = parse_arguments()
 
-    experiment = Experiment(gtsam.findExampleDataFile(args.data_file))
+    experiment = Experiment(gtsam.findExampleDataFile(args.data_file),
+                            max_loop_count=args.max_loop_count,
+                            update_frequency=args.update_frequency,
+                            max_num_hypotheses=args.max_num_hypotheses)
     experiment.run()
 
 

--- a/python/gtsam/examples/HybridCity10000.py
+++ b/python/gtsam/examples/HybridCity10000.py
@@ -121,9 +121,9 @@ def plot_estimates(gt,
             color=estimate_color,
             label=estimate_label)
     ax.legend()
-    fig.text(0.3, 0.03, text)
+    fig.text(0.1, 0.03, text)
 
-    filename = f"city10000_{text.replace('_', ' ')}.svg"
+    filename = f"city10000_{text.replace(' ', '_')}.svg"
     fig.savefig(filename, format="svg")
 
 
@@ -133,7 +133,7 @@ class Experiment:
     def __init__(self,
                  filename: str,
                  marginal_threshold: float = 1.9999,
-                 max_loop_count: int = 100,
+                 max_loop_count: int = 150,
                  update_frequency: int = 3,
                  max_num_hypotheses: int = 10,
                  relinearization_frequency: int = 10):
@@ -333,7 +333,6 @@ class Experiment:
         all_assignments = gtsam.cartesianProduct(discrete_keys)
 
         for idx, assignment in enumerate(all_assignments):
-            print(idx)
             result = gtsam.Values()
             delta = self.smoother_.hybridBayesNet().optimize(assignment)
             result.insert_or_assign(self.initial_.retract(delta))
@@ -344,7 +343,7 @@ class Experiment:
                 poses.append((pose.x(), pose.y(), pose.theta()))
             poses = np.asarray(poses)
 
-            assignment_string = "_".join([
+            assignment_string = " ".join([
                 f"{gtsam.DefaultKeyFormatter(k)}={v}"
                 for k, v in assignment.items()
             ])

--- a/python/gtsam/examples/HybridCity10000.py
+++ b/python/gtsam/examples/HybridCity10000.py
@@ -15,6 +15,7 @@ import time
 
 import numpy as np
 from gtsam.symbol_shorthand import L, M, X
+from matplotlib import pyplot as plt
 
 import gtsam
 from gtsam import (BetweenFactorPose2, HybridNonlinearFactor,
@@ -87,13 +88,52 @@ class City10000Dataset:
             return None, None
 
 
+def plot_estimates(gt,
+                   estimates,
+                   fignum: int,
+                   estimate_color=(0.1, 0.1, 0.9, 0.4),
+                   estimate_label="Hybrid Factor Graphs",
+                   text="graph"):
+    """Plot the City10000 estimates against the ground truth.
+
+    Args:
+        estimates (np.ndarray): The estimates trajectory as xy values.
+        fignum (int): The figure number for multiple plots.
+        estimate_color (tuple, optional): The color to use for the graph of estimates.
+            Defaults to (0.1, 0.1, 0.9, 0.4).
+        estimate_label (str, optional): Label for the estimates, used in the legend.
+            Defaults to "Hybrid Factor Graphs".
+    """
+    fig = plt.figure(fignum)
+    ax = fig.gca()
+    ax.axis('equal')
+    ax.axis((-65.0, 65.0, -75.0, 60.0))
+    ax.plot(gt[:, 0],
+            gt[:, 1],
+            '--',
+            linewidth=1,
+            color=(0.1, 0.7, 0.1, 0.5),
+            label="Ground Truth")
+    ax.plot(estimates[:, 0],
+            estimates[:, 1],
+            '-',
+            linewidth=1,
+            color=estimate_color,
+            label=estimate_label)
+    ax.legend()
+    fig.text(0.3, 0.03, text)
+
+    filename = f"city10000_{text.replace('_', ' ')}.svg"
+    fig.savefig(filename, format="svg")
+
+
 class Experiment:
     """Experiment Class"""
 
     def __init__(self,
                  filename: str,
-                 marginal_threshold: float = 0.9999,
-                 max_loop_count: int = 8000,
+                 marginal_threshold: float = 1.9999,
+                 max_loop_count: int = 100,
                  update_frequency: int = 3,
                  max_num_hypotheses: int = 10,
                  relinearization_frequency: int = 10):
@@ -200,7 +240,7 @@ class Experiment:
             num_measurements = len(pose_array)
 
             # Take the first one as the initial estimate
-            odom_pose = pose_array[0]
+            odom_pose = pose_array[np.random.choice(num_measurements)]
             if key_s == key_t - 1:
                 # Odometry factor
                 if num_measurements > 1:
@@ -271,7 +311,48 @@ class Experiment:
         total_time = end_time - start_time
         print(f"Total time: {total_time} seconds")
 
-        self.save_results(result, key_t + 1, time_list)
+        # self.save_results(result, key_t + 1, time_list)
+
+        # Get all the discrete values
+        discrete_keys = gtsam.DiscreteKeys()
+        for key in delta.discrete().keys():
+            # TODO Get cardinality from DiscreteFactor
+            discrete_keys.push_back((key, 2))
+        print("plotting all hypotheses")
+        self.plot_all_hypotheses(discrete_keys, key_t + 1)
+
+    def plot_all_hypotheses(self, discrete_keys, num_poses):
+        """Plot all possible hypotheses."""
+
+        # Get ground truth
+        gt = np.loadtxt(gtsam.findExampleDataFile("ISAM2_GT_city10000.txt"),
+                        delimiter=" ")
+
+        # print(discrete_keys)
+        # Get all possible assignments
+        all_assignments = gtsam.cartesianProduct(discrete_keys)
+
+        for idx, assignment in enumerate(all_assignments):
+            print(idx)
+            result = gtsam.Values()
+            delta = self.smoother_.hybridBayesNet().optimize(assignment)
+            result.insert_or_assign(self.initial_.retract(delta))
+
+            poses = []
+            for i in range(num_poses):
+                pose = result.atPose2(X(i))
+                poses.append((pose.x(), pose.y(), pose.theta()))
+            poses = np.asarray(poses)
+
+            assignment_string = "_".join([
+                f"{gtsam.DefaultKeyFormatter(k)}={v}"
+                for k, v in assignment.items()
+            ])
+
+            plot_estimates(gt,
+                           estimates=poses,
+                           fignum=idx,
+                           text=assignment_string)
 
     def save_results(self, result, final_key, time_list):
         """Save results to file."""

--- a/python/gtsam/examples/HybridCity10000.py
+++ b/python/gtsam/examples/HybridCity10000.py
@@ -112,45 +112,44 @@ class City10000Dataset:
             return None, None
 
 
-def plot_estimates(gt,
-                   estimates,
-                   fignum: int,
-                   estimate_color=(0.1, 0.1, 0.9, 0.4),
-                   estimate_label="Hybrid Factor Graphs",
-                   text="graph"):
+def plot_all_results(ground_truth,
+                     all_results,
+                     estimate_color=(0.1, 0.1, 0.9, 0.4),
+                     estimate_label="Hybrid Factor Graphs"):
     """Plot the City10000 estimates against the ground truth.
 
     Args:
-        estimates (np.ndarray): The estimates trajectory as xy values.
-        fignum (int): The figure number for multiple plots.
+        ground_truth: The ground truth trajectory as xy values.
+        all_results (List[Tuple(np.ndarray, str)]): All the estimates trajectory as xy values, as well as assginment strings.
         estimate_color (tuple, optional): The color to use for the graph of estimates.
             Defaults to (0.1, 0.1, 0.9, 0.4).
         estimate_label (str, optional): Label for the estimates, used in the legend.
             Defaults to "Hybrid Factor Graphs".
     """
-    fig = plt.figure(fignum)
-    ax = fig.gca()
-    ax.axis('equal')
-    ax.axis((-75.0, 75.0, -75.0, 75.0))
+    print(len(all_results))
+    fig, axes = plt.subplots(int(np.ceil(len(all_results) / 2)), 2)
+    for i, (estimates, text) in enumerate(all_results):
+        ax = axes[i]
+        ax.axis('equal')
+        ax.axis((-75.0, 75.0, -75.0, 75.0))
 
-    gt = gt[:estimates.shape[0]]
-    ax.plot(gt[:, 0],
-            gt[:, 1],
-            '--',
-            linewidth=1,
-            color=(0.1, 0.7, 0.1, 0.5),
-            label="Ground Truth")
-    ax.plot(estimates[:, 0],
-            estimates[:, 1],
-            '-',
-            linewidth=1,
-            color=estimate_color,
-            label=estimate_label)
-    ax.legend()
-    fig.text(0.1, 0.03, text)
+        gt = ground_truth[:estimates.shape[0]]
+        ax.plot(gt[:, 0],
+                gt[:, 1],
+                '--',
+                linewidth=1,
+                color=(0.1, 0.7, 0.1, 0.5),
+                label="Ground Truth")
+        ax.plot(estimates[:, 0],
+                estimates[:, 1],
+                '-',
+                linewidth=1,
+                color=estimate_color,
+                label=estimate_label)
+        ax.legend()
+        ax.text(0.0, 100.0, text)
 
-    filename = f"city10000_{text.replace(' ', '_')}.svg"
-    fig.savefig(filename, format="svg")
+    fig.savefig("city10000_results.svg", format="svg")
 
 
 class Experiment:
@@ -269,7 +268,8 @@ class Experiment:
             num_measurements = len(pose_array)
 
             # Take the first one as the initial estimate
-            odom_pose = pose_array[np.random.choice(num_measurements)]
+            # odom_pose = pose_array[np.random.choice(num_measurements)]
+            odom_pose = pose_array[0]
             if key_s == key_t - 1:
                 # Odometry factor
                 if num_measurements > 1:
@@ -358,14 +358,16 @@ class Experiment:
         gt = np.loadtxt(gtsam.findExampleDataFile("ISAM2_GT_city10000.txt"),
                         delimiter=" ")
 
-        # Get all possible assignments
-        if discrete_keys.size() > 5:
-            print("Too many discrete keys to plot all hypotheses. Exiting.")
-            exit(0)
+        dkeys = gtsam.DiscreteKeys()
+        for i in range(discrete_keys.size()):
+            key, cardinality = discrete_keys.at(i)
+            if key not in self.smoother_.fixedValues().keys():
+                dkeys.push_back((key, cardinality))
 
-        all_assignments = gtsam.cartesianProduct(discrete_keys)
+        all_assignments = gtsam.cartesianProduct(dkeys)
 
-        for idx, assignment in enumerate(all_assignments):
+        all_results = []
+        for assignment in all_assignments:
             result = gtsam.Values()
             gbn = self.smoother_.hybridBayesNet().choose(assignment)
 
@@ -381,21 +383,18 @@ class Experiment:
             delta = self.smoother_.hybridBayesNet().optimize(assignment)
             result.insert_or_assign(self.initial_.retract(delta))
 
-            poses = []
+            poses = np.zeros((num_poses, 3))
             for i in range(num_poses):
                 pose = result.atPose2(X(i))
-                poses.append((pose.x(), pose.y(), pose.theta()))
-            poses = np.asarray(poses)
+                poses[i] = np.asarray((pose.x(), pose.y(), pose.theta()))
 
             assignment_string = " ".join([
                 f"{gtsam.DefaultKeyFormatter(k)}={v}"
                 for k, v in assignment.items()
             ])
+            all_results.append((poses, assignment_string))
 
-            plot_estimates(gt,
-                           estimates=poses,
-                           fignum=idx,
-                           text=assignment_string)
+        plot_all_results(gt, all_results)
 
     def save_results(self, result, final_key, time_list):
         """Save results to file."""

